### PR TITLE
Modify rule S5421: Update RSPEC and add exception for volatile

### PR DESCRIPTION
**Pull request by marco-antognini-sonarsource**
## Review

A dedicated reviewer checked the rule description successfully for:

- [ ] logical errors and incorrect information
- [ ] information gaps and missing content
- [ ] text style and tone
- [ ] PR summary and labels follow [the guidelines](https://github.com/SonarSource/rspec/#to-modify-an-existing-rule)

### CODE REVIEW

File: rules/S5421/cfamily/rule.adoc [hunk 1]
```diff
@@ -1,15 +1,21 @@
 == Why is this an issue?
 
-A global variable can be modified from anywhere in the program. At first, this might look convenient, but it makes programs harder to understand. When you see a function call, you cannot know if the function will affect the value of the variable or not. You have lost the ability to reason locally about your code and must always have the whole program in mind.
+A global variable can be modified from anywhere in the program. At first, this might look convenient. However, it makes programs harder to understand and maintain. When you see a function call, you cannot know if the function will affect the value of the global variable or not. You have lost the ability to reason locally about your code and must always have the whole program in mind.
 
 
-Additionally, global variables are often subject to race conditions in multi-threaded environments.
+Additionally, global variables are often subject to race conditions in multi-threaded environments. To limit risks, global pointers should be `const` at every level:
```
Comment (loic-joly-sonarsource) [inline rules/S5421/cfamily/rule.adoc:6]: By writing it this way, it appears that being const only addresses the second problem, while it also addresses the first one.
```suggestion
Additionally, global variables are often subject to race conditions in multi-threaded environments.

These issues are related to modification and cannot occur when the global variable is `const` (or, in the case of a pointer, if it is `const` at every level).
```

File: rules/S5421/cfamily/rule.adoc [hunk 2]
```diff
@@ -1,15 +1,21 @@
...
+unsigned** noncompliantPtr;
+unsigned const* const* const compliantPtr = ...;
+----
 
 Some global variables defined in external libraries (such as ``++std::cout++``, ``++std::cin++``, ``++std::cerr++``) are acceptable to use, but you should have a good reason to create your own. If you use a global variable, ensure they can be safely accessed concurrently.
 
+Remember that it is much easier to maintain software without globals. Instead of such variables, it is better to design functions to take as input all the required variables. In addition to serving documentation, this also helps future refactoring and the evolution of the code.
 
-This rule detects all declarations of global variables (in the global namespace or any namespace) that are not constant.
+This rule detects all declarations of global variables (in any namespace) that are not constant.
```
Comment (loic-joly-sonarsource) [inline rules/S5421/cfamily/rule.adoc:18]: For some people, the global namespace is not really a namespace... 
```suggestion
This rule detects all declarations of global variables (at file scope or in any namespace) that are not constant.
```

File: rules/S5421/cfamily/rule.adoc [hunk 3]
```diff
@@ -34,15 +41,29 @@ int main() {
...
+  auto userValue = readValueInFeet();
   writeResult(userValue * footToMeter);
 }
 ----
 
+=== Exceptions
+
+In embedded software, `volatile` is often used to indicate that some piece of memory can be mutated by external factors even in single-threaded applications.
+
+It remains good practice to have `const` global pointers when these global variables cannot be avoided. That being said, it is acceptable -- and often required, but not always -- that the pointee is a non-`const` `volatile` type:
```
Comment (loic-joly-sonarsource) [inline rules/S5421/cfamily/rule.adoc:57]:
```suggestion
`volatile` is used to indicate that some piece of memory can be mutated by external factors. For embedded software, some hardware inputs/outputs can be mapped to specific memory addresses, and accessing these bound data is usually done through a global pointer to `volatile` data.

In that situation, the pointer itself should be const, but the pointee can be non-const if the memory maps an output register that is supposed to be written to.
```

File: rules/S5421/cfamily/rule.adoc [hunk 4]
```diff
@@ -34,15 +41,29 @@ int main() {
...
+
+[source,cpp]
+----
+unsigned volatile*       gpio;       // Noncompliant
+unsigned volatile* const gpio = ...; // Compliant
+
+// When possible, prefer `const` at every level:
+unsigned volatile const* const readOnlyHardwareRegister = ...;
```
Comment (loic-joly-sonarsource) [inline rules/S5421/cfamily/rule.adoc:65]:
```suggestion
unsigned volatile* const gpio2 = ...; // Compliant, used for input and output
unsigned volatile const* const gpio3 = ...; // Compliant, used for input only

```

Comment (marco-antognini-sonarsource) [reply rules/S5421/cfamily/rule.adoc]: Slightly adapted in 426a963b29

**Review by loic-joly-sonarsource — APPROVED** [review]

**Pull request merged**